python/pyspark/sql/observation.py

```diff
@@ -0,0 +1,146 @@
+    function.
+
+    Example:
+        >>> from pyspark.sql.functions import col, count, lit, max
+        >>> from pyspark.sql.observation import Observation
+        >>> observation = Observation("my_metrics")
+        >>> observed_df = df.observe(observation, count(lit(1)), max(col("age")))
+        >>> observed_df.count()
+        >>> metrics = observation.get
+
+    An Observation instance collects the metrics while the first action is executed. Subsequent
+    actions do not modify the metrics returned by `Observation.get`. Retrieval of the metric via
+    `Observation.get` blocks until the first action has finished and metrics become available.
+
+    This class does not support streaming datasets.
+    """
+    def __init__(self, name: Optional[str] = None):
+        """Constructs a named or unnamed Observation instance.
+
+        parameters
+        ----------
+        name : str, optional
+            default is a random UUID string. This is the name of the Observation and the metric.
+        """
+        assert isinstance(name, basestring), "name should be a string"
+        self._name = name or str(uuid4())
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df: DataFrame, *exprs: Column) -> DataFrame:
+        """Attaches this observation to the given :class:`DataFrame` to observe aggregations.
+
+        parameters
+        ----------
+        df : :class:`DataFrame`
+            the :class:`DataFrame` to be observed
+        exprs : list of :class:`Column`
+            column expressions (:class:`Column`).
+
+        returns
+        -------
+        :class:`DataFrame`
+            the observed :class:`DataFrame`.
+        """
+        assert exprs, "exprs should not be empty"
+        assert all(isinstance(c, Column) for c in exprs), "all exprs should be Column"
+        assert self._jo is None, "an Observation can be used with a DataFrame only once"
+
+        self._jvm = df._sc._jvm
+        self._jo = self._jvm.org.apache.spark.sql.Observation(self._name)
+        observed_df = self._jo.on(df._jdf,
+                                  exprs[0]._jc,
+                                  column._to_seq(df._sc, [c._jc for c in exprs[1:]]))
+        return DataFrame(observed_df, df.sql_ctx)
+
+    @property
+    def get(self) -> Row:
+        """Get the observed metrics.
+
+        Waits until the observed dataset finishes its first action. Only the result of the
+        first action is available. Subsequent actions do not modify the result.
+
+        returns
+        -------
+        :class:`Row`
+            the observed metrics
+        """
+        assert self._jo is not None, 'call DataFrame.observe'
+        jrow = self._jo.get()
+        return self._to_row(jrow)
+
+    def _to_row(self, jrow):
+        field_names = jrow.schema().fieldNames()
+        valuesScalaMap = jrow.getValuesMap(self._jvm.PythonUtils.toSeq(list(field_names)))
```

Comment:
Let's keep them in the under_score_naming whenever possible (except calling Py4J and user-facing API name)

Reply:
fixed